### DISCUSSION TITLE
[RFC] A Contributors Static Site

### DISCUSSION THREAD

**Discussion by indifferentghost** [contributers-ssg/contributors-rfc#1] [Ideas]
# Abstract

This document describes the design and implementation of "Contributors", a statically generated website using Astro and GitHub Pages. The application will query the GitHub API using Flat Data, fetching user contribution data from multiple repositories, and rendering the data in a user-friendly manner.

# Background

Open-source projects and repositories rely heavily on the contributions of developers worldwide. Recognizing and displaying the contributions of these individuals is essential for motivating and appreciating their efforts. This proposal aims to create a simple and efficient way to display contributor data for multiple repositories.

# Proposal
## Overview

We propose the development of the "Contributors" application, a statically generated website built using Vite and deployed on GitHub Pages. The application will utilize Flat Data to query the GitHub API, fetching contribution data for specified repositories, and displaying this information in a visually appealing and organized manner.

### Vite

[Vite](https://vitejs.dev/) will be used as the build tool and development server for this project. It offers fast development, efficient production build generation, and support for popular JavaScript frameworks like React, which is required for Material-UI.

### GitHub Pages

[GitHub Pages](https://pages.github.com/) will serve as the hosting platform for this project. It provides an easy-to-use, free, and reliable solution for hosting static websites directly from a GitHub repository. Additionally, GitHub Pages offers built-in CI/CD, allowing for continuous deployment of the application.

### Flat Data

[Flat Data](https://githubnext.com/projects/flat-data) will be used for fetching contribution data from the GitHub API. Flat Data is a GitHub Action that fetches data from any API or URL and stores the data in a repository as a flat file. It is a convenient and efficient solution for fetching and storing the required data for this project.
Data Fetching and Processing

The application will use Flat Data to query the GitHub API for the following information:
- Repository data for specified repositories.
- User contribution data for the specified repositories, including the user's avatar, name, username, and the number of contributions made.

### Material-UI

[Material-UI](https://mui.com/) is a popular React UI framework that implements Google's Material Design guidelines. It offers a comprehensive set of pre-built components that are customizable, responsive, and accessible, making it an excellent choice for the "Contributors" application.

## User Interface

The UI will display the repository data and the user contribution data in a visually appealing and organized manner. The application will include:

- header displaying the name of the project, using the `AppBar` and `Toolbar` components.
- A list of repositories with their name, description, and link to their GitHub page, using the `Card`, `CardContent`, and `Typography` components.
- An expandable section under each repository, displaying a list of contributors with their avatar, name, username, and the number of contributions made, using the `Accordion`, `AccordionSummary`, `AccordionDetails`, `List`, and `ListItem` components.
- Pagination, if necessary, to accommodate a large number of repositories or contributors, using the `Pagination` component.

# Conclusion

The "Contributors" application provides a simple and efficient way to display contributor data for multiple repositories, recognizing and appreciating the efforts of developers worldwide. Using Astro, GitHub Pages, and Flat Data, this application will deliver a fast, optimized, and visually appealing solution for showcasing contributions to open-source projects.

**Comment by indifferentghost** [comment]
**NOTE** This was written heavily by GPT-4, if there's problems please feel free to comment.

**Comment by indifferentghost** [comment]
I think the functionality listed below is excellent for an MVP
> - A list of repositories with their name, description, and link to their GitHub page, using the `Card`, `CardContent`, and `Typography` components.
> - An expandable section under each repository, displaying a list of contributors with their avatar, name, username, and the number of contributions made, using the `Accordion`, `AccordionSummary`, `AccordionDetails`, `List`, and `ListItem` components.
> - Pagination, if necessary, to accommodate a large number of repositories or contributors, using the `Pagination` component.

**Reply by indifferentghost** [reply]
Though there's more components that could be created here, It might be worth doing a deeper explanation as I know the [`avatar`](https://mui.com/material-ui/react-avatar/) element exists.

**Reply by indifferentghost** [reply]
I've created issue #2 to take care of this spike.

**Reply by dev-skworks** [reply]
Thank you!! The deeper explanation is definitely helpful.

Also, what's a spike in this context?

**Comment by indifferentghost** [comment]
Shoot it seems like [Astro doesn't have great support for CSS-in-JS yet](https://github.com/withastro/astro/issues/4432), which is required by Material-UI in the form of emotion. Falling back to [vite](https://vitejs.dev/). It's the core of astro and does support CSS-in-JS

**Reply by dev-skworks** [reply]
Thank you for discovering this and making the change! Also for updating the Technical Details section.

The Abstract and the Conclusion sections referenced below could also use an update since we aren't using Astro anymore.

> # Abstract
> 
> This document describes the design and implementation of "Contributors", a statically generated website using Astro and GitHub Pages. The application will query the GitHub API using Flat Data, fetching user contribution data from multiple repositories, and rendering the data in a user-friendly manner.
>
> # Conclusion
> 
> The "Contributors" application provides a simple and efficient way to display contributor data for multiple repositories, recognizing and appreciating the efforts of developers worldwide. Using Astro, GitHub Pages, and Flat Data, this application will deliver a fast, optimized, and visually appealing solution for showcasing contributions to open-source projects.

**Comment by dev-skworks** [comment]
> * An expandable section under each repository, displaying a list of contributors with their avatar, name, username, and the number of contributions made, using the `Accordion`, `AccordionSummary`, `AccordionDetails`, `List`, and `ListItem` components.
> 
> * Pagination, if necessary, to accommodate a large number of repositories or contributors, using the `Pagination` component.

So from what I am understanding, the `Pagination` component will exist inside the expandable section? Specifically, inside the `AccordionDetails` component?

I haven't used/researched MUI much so I want to make sure that I am understanding the MVP requirements correctly :)
